HIP/hip-906.md

```diff
@@ -0,0 +1,125 @@
+
+
+## User stories
+
+1. As a smart contract developer, I want to be able to grant an allowance for hbars from an EOA to a spender account or contract.
+2. As a smart contract developer, I want to be able to grant an allowance for hbars from a contract to a spender account or contract.
+3. As a smart contract developer, I want to be able to get the current allowance of hbars from an owner account to a spender account or contract.
+
+## Specification
+
+HIP-632 proposes to introduce a new system contract for accessing Hedera account functionality (Hedera Account Service - HAS).
+This HIP extends the functionality of the new system contract by adding another related interface `IHRC632` to support the `hbarAllowance` and `hbarApprove` functions.
+The `hbarAllowance` function will be used to retrieve information about allowance granted to a spender and the `hbarApprove` function will allow the sender to grant to the `spender` an allowance of hbars.
+
+
+Similar to the way redirection to a proxy contract during EVM execution for tokens works [see HIP-719](https://github.com/hashgraph/hedera-improvement-proposal/blob/main/HIP/hip-719.md),
+this HIP proposes to introduce a new proxy contract for accounts.  During EVM execution, if the target of the current frame is an account, a call to the new proxy contract will be created and the current calldata will be injected into 
+the proxy contract for processing by the Hedera Account Service system contract.
+
+The bytecode for the proxy contract can be created by compiling the following contract:
+
+```solidity
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+contract Assembly {
+	fallback() external {
+		address precompileAddress = address(0x16a);
+		assembly {
+			mstore(0, 0xFEFEFEFEFEFEFEFEFEFEFEFEFEFEFEFEFEFEFEFEFEFEFEFE)
+			calldatacopy(32, 0, calldatasize())
+			let result := delegatecall(gas(), precompileAddress, 8, add(24, calldatasize()), 0, 0)
+			let size := returndatasize()
+			returndatacopy(0, 0, size)
+			switch result
+				case 0 { revert(0, size) }
+				default { return(0, size) }
+		}
+	}
+}
+```
+
+The following table describes the function selector for the new `hbarAllownace` and `hbarApprove` functions and the associated function signature and response.
+
+| Function Selector Hash | Function Signature                          | Response                                                                  | 
+|------------------------|---------------------------------------------|---------------------------------------------------------------------------|
+| 0xbbee989e             | hbarAllowance(address spender)              | (ResponseCode, int256 - amount of hbar allowances granted to the spender) | 
+| 0x86aff07c             | hbarApprove(spender address, amount int256) | ResponseCode                                                              |
+
+The solidity interface for IHRC632 will be the following
+
+```
+interface IHRC632 {
```

Comment:
Add a call out to note that like other system contracts delegate call of these functions is not supported by the network

Reply:
done